### ISSUE TITLE
[BUG] HttpMessage.Properties not accessible anymore after dispose.

### ISSUE THREAD

**Issue by Socolin** [Azure/azure-sdk-for-net#34889]
### Library name and version

Azure.Core 1.30.0

### Describe the bug

Hello,

This commit https://github.com/Azure/azure-sdk-for-net/commit/437d4ad07ed5580aa8aac690701c05a62d18049b broke our usage of `HttpPipeline`  in our tests:

We use a fake transport to catch all message sent through the HttpPipeline, then in the asserts we check if the received message are what we are expecting.

With this changes the `_propertyBag` is disposed and then it's impossible to access the properties anymore.

### Expected behavior

Property should be accessible or we should have a way to copy all of them in the transport layer, maybe with a way to list all keys so we can make a copy of it.

### Actual behavior

Property are note available anymore

### Reproduction Steps

Capture message with an `HttpPipelineTransport` and try to access the properties after the `Dispose`

```cs
public class FakeTransport : HttpPipelineTransport
{
	private readonly List<FakeRequest> _requests = new List<FakeRequest>();
	private readonly List<HttpMessage> _receivedMessages = new List<HttpMessage>();
	public IReadOnlyList<HttpMessage> ReceivedMessages => _receivedMessages;
	public IReadOnlyList<FakeRequest> CreatedRequests => _requests;
	public Action<HttpMessage>? OnProcess { get; set; }

	public FakeTransport()
	{
		OnProcess = httpMessage =>
			httpMessage.Response = new FakeResponse(418, "OnProcess must be override in tests that directly need to control the response ");
	}

	public override void Process(HttpMessage message)
	{
		OnProcess?.Invoke(message);
		_receivedMessages.Add(message);
	}

	public override ValueTask ProcessAsync(HttpMessage message)
	{
		OnProcess?.Invoke(message);
		_receivedMessages.Add(message);
		return new ValueTask();
	}

	public override Request CreateRequest()
	{
		var request = new FakeRequest();
		_requests.Add(request);
		return request;
	}
}
```

**Comment by jsquire** [comment]
//cc: @annelo-msft, @JoshLove-msft, @AlexanderSher

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by christothes** [comment]
Hi @Socolin -
Could you clarify how the message instances are getting disposed in your tests?

**Comment by Socolin** [comment]
They are disposed in our production code, for example:

```cs
public async Task<SomethingResponse> GetSomething(
	CancellationToken cancellationToken = default
)
{
	using var message = new HttpMessageBuilder(_pipeline)
		.WithMethod(RequestMethod.Get)
		.WithPath($"/some-path")
		.Build();

	await _pipeline.SendAsync(message, cancellationToken).ConfigureAwait(false);

	return await ParseMessageAndHandleCommonErrorAsync<SomethingResponse>(message, cancellationToken: cancellationToken).ConfigureAwait(false);
}
```

and HttpMessageBuilder.Build() is just creating the message with 

```cs
var message = _pipeline.CreateMessage();;

...

return message
```

So in the test when we call ` var actual = await client.GetSomething()` and then check the message recieved by the `HttpTransport`  it has been disposed.

I don't think the dispose is a problem, I'm just looking for a way to capture all the properties in the Transport layer to be able to check it in the asserts

**Comment by christothes** [comment]
The way we handle this in our tests is to perform the asserts in the [request handler of the mock transport](https://github.com/Azure/azure-sdk-for-net/blob/f838465a065520c3ea478d5fee254a7eefb1be59/sdk/core/Azure.Core.TestFramework/src/MockTransport.cs#L45-L48). Below is an example:

https://github.com/Azure/azure-sdk-for-net/blob/f838465a065520c3ea478d5fee254a7eefb1be59/sdk/identity/Azure.Identity/tests/AuthorizationCodeCredentialTests.cs#L94-L102

**Comment by Socolin** [comment]
I though about changing for this, but this make the test less readable, I always try to keep the 
```
// arrange
// act
// assert
```

With  the suggested change the assert is mixed in the `arrange` sections and it make the test harder to quickly figure out what is tested, I'm really attached to keeping the test readable with same quality as production code.

In those test we also test the method, header, query strings... so having 1 part that is tested in a different way is really not ideal. 

And since we have a large number of tests that are using this, doing what you propose will take us quite some time since we cannot fix this with a find and replace, and this may lead to creating errors / typo

Is there any chance to expose a `.PropertyKeys` or something like that ?  this way I can get a copy of the properties and continue testing this the same way as before

**Comment by christothes** [comment]
> Is there any chance to expose a `.PropertyKeys` or something like that ? this way I can get a copy of the properties and continue testing this the same way as before

I don't believe so - HttpMessage is in the critical path from a perf perspective, so I don't see how we could do that without affecting allocations. As an implementation detail, the property bag is implemented as an array rented from the ArrayPool.

**Comment by christothes** [comment]
Yes, there is a dictionary, but it is only referenced inside the array, as you pointed out. Also, this is an internal implementation detail that could also change in the future (as it did in this case ). Unfortunately, the property bag is a resource owned by the type that must be cleaned up on disposal.

**Comment by Socolin** [comment]
I re-read the code and I saw the overload `SetProperty(Type type, object value)`

So if I understand correctly, the `Dictionary` is kind of the "slow" way with string and for performance we should use like a class, and store the property as an object.

So `Keys` does not really exists, there are 2 kind of keys and since the `TypeHandle` is stored in one case I think it would be complex to get the list back.

I found a simple way to fix my problem and keep the test as I want them (arrange/act/assert).

Thanks for the explanations.
